tachyon/base/containers/adapters.h

```diff
@@ -206,6 +206,92 @@ class ZippedAdapter {
   U& u_;
 };
 
+// Internal adapter class for implementing tachyon::base::Chained.
+template <typename T>
+class ChainedAdapter {
+ public:
+  class Iterator {
+   public:
+    using ItType = decltype(std::begin(std::declval<T&>()));
+
+    using difference_type = std::ptrdiff_t;
+    using value_type =
+        std::conditional_t<std::is_const_v<T>, const iter_value_t<ItType>,
+                           iter_value_t<ItType>>;
+    using pointer = value_type*;
+    using reference = value_type&;
+    using iterator_category = std::forward_iterator_tag;
+
+    static Iterator begin(T& first, T& second) {
+      return Iterator(std::begin(first), std::end(first), std::begin(second),
+                      std::end(second), true);
+    }
+    static Iterator end(T& first, T& second) {
+      return Iterator(std::end(second), std::end(first), std::begin(second),
+                      std::end(second), false);
+    }
+
+    Iterator(const Iterator& other) = default;
+    Iterator& operator=(const Iterator& other) = default;
+
+    bool operator==(const Iterator& other) const { return it_ == other.it_; }
+    bool operator!=(const Iterator& other) const { return !(*this == other); }
+
+    Iterator& operator++() {
+      if (is_first_) {
+        it_++;
+        if (it_ == end_first_) {
+          it_ = begin_second_;
+          is_first_ = false;
+        }
+      } else {
+        if (it_ != end_second_) {
+          ++it_;
+        }
+      }
+      return *this;
+    }
+
+    Iterator operator++(int) {
+      Iterator it(*this);
+      ++(*this);
+      return it;
+    }
+
+    pointer operator->() { return it_; }
```

Comment:
Please refer to other implementations to overcome `-Werror,-Wignored-reference-qualifiers`

Plus,  the type `pointer` isn't matched to `iterator`, to match it,

```suggestion
    pointer operator->() { return &*it_; }
```